File: ac/abc/149/d.cpp
#include<bits/stdc++.h>
using namespace std;
typedef long long ll;

int main(){
	ll n, k, r, s, p;
	scanf("%lld %lld %lld %lld %lld", &n, &k, &r, &s, &p);
	string t; cin>>t;

	vector<string> ss(k);
	for(ll i=0; i<k; i++) ss[i] = "";
	for(ll i=0; i<n; i++) ss[i%k] += t[i];

	ll ans = 0;
	for(ll i=0; i<k; i++){
		auto prev = ss[i][0];
		ll cnt = 0;
		for(auto c: ss[i]){
			if(c==prev) cnt++;
			else{
				ll tmp = (cnt+1)/2;
				if(prev=='r') ans += tmp*p;
				else if(prev=='s') ans += tmp*r;
				else ans += tmp*s;
				cnt = 1;
			}
			prev = c;
		}
		ll tmp = (cnt+1)/2;
		if(prev=='r') ans += tmp*p;
		else if(prev=='s') ans += tmp*r;
		else ans += tmp*s;
	}
	printf("%lld\n", ans);
	return 0;
}

File: ac/abc/149/d.nim
import strutils, sequtils, future, math
var
  n, k, ro, sc, pa: int
  t: string
(n, k) = stdin.readline.split.map(parseInt)
(ro, sc, pa) = stdin.readline.split.map(parseInt)
t = stdin.readline

var ss = lc["" | (x<-0..<k), string]
for i in 0..<n: ss[i mod k] = ss[i mod k] & t[i]

var ans = 0
for s in ss:
  var
    prev=s[0]
    cnt=0

  for c in s:
    if prev==c:
      cnt += 1
    else:
      var tmp = (cnt+1) div 2
      case prev
        of 'r': ans += tmp * pa
        of 's': ans += tmp * ro
        of 'p': ans += tmp * sc
        else: discard
      cnt = 1
    prev = c

  var tmp = (cnt+1) div 2
  case prev
    of 'r': ans += tmp * pa
    of 's': ans += tmp * ro
    of 'p': ans += tmp * sc
    else: discard
echo ans
